lib/Jsrt/JsrtDiag.cpp

```diff
@@ -0,0 +1,688 @@
+
+        *properties = JS_INVALID_REFERENCE;
+
+        JsrtContext* context = JsrtContext::GetCurrent();
+        JsrtRuntime* runtime = context->GetRuntime();
+
+        VALIDATE_RUNTIME_IS_AT_BREAK(runtime);
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        JsrtDebuggerStackFrame* debuggerStackFrame = nullptr;
+        if (!jsrtDebugManager->TryGetFrameObjectFromFrameIndex(scriptContext, stackFrameIndex, &debuggerStackFrame))
+        {
+            return JsErrorDiagObjectNotFound;
+        }
+
+        Js::DynamicObject* localsObject = debuggerStackFrame->GetLocalsObject(scriptContext);
+
+        if (localsObject != nullptr)
+        {
+            *properties = localsObject;
+            return JsNoError;
+        }
+
+        return JsErrorDiagUnableToPerformAction;
+    });
+}
+
+CHAKRA_API JsDiagGetProperties(
+    _In_ unsigned int objectHandle,
+    _In_ unsigned int fromCount,
+    _In_ unsigned int totalCount,
+    _Out_ JsValueRef *propertiesObject)
+{
+
+    return ContextAPIWrapper<false>([&](Js::ScriptContext *scriptContext) -> JsErrorCode {
+
+        PARAM_NOT_NULL(propertiesObject);
+
+        *propertiesObject = JS_INVALID_REFERENCE;
+
+        JsrtContext* context = JsrtContext::GetCurrent();
+        JsrtRuntime* runtime = context->GetRuntime();
+
+        VALIDATE_RUNTIME_IS_AT_BREAK(runtime);
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        JsrtDebuggerObjectBase* debuggerObject = nullptr;
+        if (!jsrtDebugManager->GetDebuggerObjectsManager()->TryGetDebuggerObjectFromHandle(objectHandle, &debuggerObject) || debuggerObject == nullptr)
+        {
+            return JsErrorDiagInvalidHandle;
+        }
+
+        Js::DynamicObject* properties = debuggerObject->GetChildrens(scriptContext, fromCount, totalCount);
+
+        if (properties != nullptr)
+        {
+            *propertiesObject = properties;
+            return JsNoError;
+        }
+
+        return JsErrorDiagUnableToPerformAction;
+    });
+}
+
+CHAKRA_API JsDiagGetObjectFromHandle(
+    _In_ unsigned int objectHandle,
+    _Out_ JsValueRef *handleObject)
+{
+    return ContextAPIWrapper<false>([&](Js::ScriptContext *scriptContext) -> JsErrorCode {
+
+        PARAM_NOT_NULL(handleObject);
+
+        *handleObject = JS_INVALID_REFERENCE;
+
+        JsrtContext* context = JsrtContext::GetCurrent();
+        JsrtRuntime* runtime = context->GetRuntime();
+
+        VALIDATE_RUNTIME_IS_AT_BREAK(runtime);
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        JsrtDebuggerObjectBase* debuggerObject = nullptr;
+        if (!jsrtDebugManager->GetDebuggerObjectsManager()->TryGetDebuggerObjectFromHandle(objectHandle, &debuggerObject) || debuggerObject == nullptr)
```

Comment:
you could create a lambda and push the common code in the lambda caller. 

---
In reply to: [62749979](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62749979) [](ancestors = 62749979)